openedx/core/djangoapps/models/course_details.py

```
@@ -71,10 +71,10 @@ def __init__(self, org, course_id, run):
         self.pre_requisite_courses = []  # pre-requisite courses
         self.entrance_exam_enabled = ""  # is entrance exam enabled
         self.entrance_exam_id = ""  # the content location for the entrance exam
-        self.entrance_exam_minimum_score_pct = settings.FEATURES.get(
+        self.entrance_exam_minimum_score_pct = str(settings.FEATURES.get(
```

Why was this change necessary?

---

Yeah, this is a problematic pattern here. Let me quickly walk you through it! 

The `ENTRANCE_EXAM_MIN_SCORE_PCT` is not actually a feature flag, it's a regular Django setting, but it's being accessed through `settings.FEATURES.get()` as if it were a feature setting(Not sure why).

What's happening here is,

Our FeatureProxy has access to the entire settings namespace, not just feature flags, When `ENTRANCE_EXAM_MIN_SCORE_PCT` (a non-feature setting) is accessed via `FeatureProxy`, it returns the actual setting value: **50** (integer). However, `self.entrance_exam_minimum_score_pct` expects a string ("50"). This type mismatch breaks CI.

It worked fine with `FEATURES` Dictionary because it doesn't contain `ENTRANCE_EXAM_MIN_SCORE_PCT`, so `settings.FEATURES.get('ENTRANCE_EXAM_MIN_SCORE_PCT', '50')` always falls back to the string default **"50"**, avoiding the type issue.

To solve this, we can either just wrap `str` around it or  use getattr like this  `str(getattr(settings, 'ENTRANCE_EXAM_MIN_SCORE_PCT', 50))`. 

Please let me your thoughts on this!

---

Oh yikes 😅 Good detective work. This is almost certainly a bug--the original dev must have mistakenly looked in FEATURES instead of accessing the attribute on `settings`. (Side note: Bugs like these are part of why we are doing this refactoring in the first place. Having toggles in a dictionary encourages developers to use `.get(...)`, which silently swallows errors when a setting is missing, whereas if you misspell an attribute of `settings`, you'll find out right away with an AttributeError).

Maybe we should just fix the bug and change it to `str(settings.ENTRACE_EXAM_MIN_SCORE_PCT)`. The default defined as in common.py is already 50, so for operators who weren't overriding the setting anyway, there will be no change in behavior. And, if an operator is overriding the setting in their `FEATURES` dictionary, then your FeaturesProxy should ensure that is reflected on the `settings` object. What do you think @wgu-ram-chandra ?

---

Thanks! I agree with your point. The `ENTRACE_EXAM_MIN_SCORE_PCT` setting is initialized in both `cms/envs/common.py` and `cms/envs/test.py`. Correct me if I’m wrong, but since the `openedx/` module defaults to `devstack.py` for both LMS and CMS (as observed from `manage.py`), and `devstack.py` already includes `ENTRACE_EXAM_MIN_SCORE_PCT`, it’s puzzling that accessing `settings.ENTRACE_EXAM_MIN_SCORE_PCT` raises an `AttributeError`.

Interestingly, when running in CMS environment, `settings.ENTRACE_EXAM_MIN_SCORE_PCT` works fine. So to avoid potential errors, it would be safer to use:

```python
getattr(settings, 'ENTRACE_EXAM_MIN_SCORE_PCT', 50)
```

This ensures a default value is used if the setting isn’t defined, preventing the application from breaking.

---

Where do you see it raising an AttributeError? In an LMS context? If so, that is most likely because lms/envs/common.py does not define ENTRANCE_EXAM_MIN_SCORE_PCT; in other words, the setting is only defined in the CMS context.

I'd rather not use `getattr`. Generally, getattr is bad because it fails quietly if its argument is misspelled or if the underlying setting changes or goes away. Instead, could you move the definition of ENTRANCE_EXAM_MIN_SCORE_PCT up to openedx/envs/common.py so that it's available in all contexts, and then just access the setting as `settings.ENTRANCE_EXAM_MIN_SCORE_PCT`?

---

Awesome! got it! I’ve moved the definition to `openedx/envs/common.py`, and `settings. ENTRANCE_EXAM_MIN_SCORE_PCT` works.